src/plugins/workspace/public/components/workspace_form/select_data_source_panel.tsx

```diff
@@ -3,149 +3,171 @@
+import { useOpenSearchDashboards } from '../../../../opensearch_dashboards_react/public';
+import { WorkspaceClient } from '../../workspace_client';
+import { AssociationDataSourceModalMode } from '../../../common/constants';
+import { DataSourceConnectionTable } from '../workspace_detail/data_source_connection_table';
+import { fetchDataSourceConnections } from '../../utils';
+import { DataSourceEngineType } from '../../../../data_source/common/data_sources';
 
 export interface SelectDataSourcePanelProps {
   errors?: { [key: number]: WorkspaceFormError };
   savedObjects: SavedObjectsStart;
-  selectedDataSources: DataSource[];
+  assignedDataSources: DataSource[];
   onChange: (value: DataSource[]) => void;
+  isDashboardAdmin: boolean;
 }
 
 export const SelectDataSourcePanel = ({
   errors,
   onChange,
-  selectedDataSources,
+  assignedDataSources,
   savedObjects,
+  isDashboardAdmin,
 }: SelectDataSourcePanelProps) => {
-  const [dataSourcesOptions, setDataSourcesOptions] = useState<EuiComboBoxOptionOption[]>([]);
+  const [modalVisible, setModalVisible] = useState(false);
+  const [selectedItems, setSelectedItems] = useState<DataSourceConnection[]>([]);
+  const [assignedDataSourceConnections, setAssignedDataSourceConnections] = useState<
+    DataSourceConnection[]
+  >([]);
+  const [toggleIdSelected, setToggleIdSelected] = useState(
+    AssociationDataSourceModalMode.OpenSearchConnections
+  );
+  const {
+    services: { notifications, http, chrome },
+  } = useOpenSearchDashboards<{ CoreStart: CoreStart; workspaceClient: WorkspaceClient }>();
+
   useEffect(() => {
-    if (!savedObjects) return;
-    getDataSourcesList(savedObjects.client, ['*']).then((result) => {
-      const options = result.map(({ title, id }) => ({
-        label: title,
-        value: id,
-      }));
-      setDataSourcesOptions(options);
+    fetchDataSourceConnections(assignedDataSources, http, notifications).then((connections) => {
+      setAssignedDataSourceConnections(connections);
     });
-  }, [savedObjects, setDataSourcesOptions]);
-  const handleAddNewOne = useCallback(() => {
-    onChange?.([
-      ...selectedDataSources,
-      {
-        title: '',
-        id: '',
-      },
-    ]);
-  }, [onChange, selectedDataSources]);
+  }, [assignedDataSources, http, notifications]);
+
+  const handleAssignDataSources = (dataSourceConnections: DataSourceConnection[]) => {
+    setModalVisible(false);
+    const dataSources = dataSourceConnections
+      .filter(
+        ({ connectionType }) => connectionType === DataSourceConnectionType.OpenSearchConnection
+      )
+      .map(({ id, type, name, description }) => ({
+        id,
+        title: name,
+        description,
+        dataSourceEngineType: type as DataSourceEngineType,
+      }));
+    const savedDataSources: DataSource[] = [...assignedDataSources, ...dataSources];
+    onChange(savedDataSources);
```

Comment:
Can we call `setAssignedDataSourceConnections` immediately in this function? The selected connections in table need to wait for fetching connections after data source change.

Reply:
Alright, maybe the `useEffect` here can be removed.
```
  useEffect(() => {
    fetchDataSourceConnections(assignedDataSources, http, notifications).then((connections) => {
      setAssignedDataSourceConnections(connections);
    });
  }, [assignedDataSources, http, notifications]);
```